tests/Elasticsearch/Metadata/Resource/Factory/ElasticsearchOperationProviderResourceMetadataCollectionFactoryTest.php

```diff
@@ -0,0 +1,96 @@
+        self::expectException(\LogicException::class);
+        $resourceMetadataFactory->create(Foo::class);
+    }
+
+    public function testWithElasticsearchDocument(): void
+    {
+        $get = new Get(name: 'foo_get', persistenceMeans: new ElasticsearchDocument());
+
+        $resourceMetadataFactory = new ElasticsearchOperationProviderResourceMetadataCollectionFactory($this->getResourceMetadataCollectionFactory(Foo::class, $get));
+        /** @var Get $operationResult */
+        $operationResult = $resourceMetadataFactory->create(Foo::class)->getOperation('foo_get');
+        self::assertSame(ElasticsearchItemProvider::class, $operationResult->getProvider());
+    }
+
+    public function testWithoutElasticsearchDocument(): void
+    {
+        $get = new Get(name: 'foo_get');
+
+        $resourceMetadataFactory = new ElasticsearchOperationProviderResourceMetadataCollectionFactory($this->getResourceMetadataCollectionFactory(Foo::class, $get));
+        $operationResult = $resourceMetadataFactory->create(Foo::class)->getOperation('foo_get');
+        self::assertNull($operationResult->getProvider());
+    }
+
+    public function testExistingProviderIsNotReplaced(): void
+    {
+        $get = new Get(name: 'foo_get', provider: 'foo', persistenceMeans: new ElasticsearchDocument());
+
+        $resourceMetadataFactory = new ElasticsearchOperationProviderResourceMetadataCollectionFactory($this->getResourceMetadataCollectionFactory(Foo::class, $get));
+        $operationResult = $resourceMetadataFactory->create(Foo::class)->getOperation('foo_get');
+        self::assertSame('foo', $operationResult->getProvider());
+    }
+
+    public function testWithGraphqlOperation(): void
+    {
+        $query = new Query(name: 'foo_query', persistenceMeans: new ElasticsearchDocument());
+
+        $resourceMetadataFactory = new ElasticsearchOperationProviderResourceMetadataCollectionFactory($this->getResourceMetadataCollectionFactory(Foo::class, $query));
+        $operationResult = $resourceMetadataFactory->create(Foo::class)->getOperation($query->getName());
+        self::assertSame(ElasticsearchItemProvider::class, $operationResult->getProvider());
+    }
+
+    /**
+     * @return Stub&ResourceMetadataCollectionFactoryInterface
+     */
+    private function getResourceMetadataCollectionFactory(string $resourceClass, Operation $operation): Stub
+    {
+        if ($operation instanceof GraphQlOperation) {
+            $resource = new ApiResource(graphQlOperations: [$operation->getName() => $operation]);
+        } else {
+            $resource = new ApiResource(operations: [$operation->getName() => $operation]);
+        }
+        $metadata = new ResourceMetadataCollection($resourceClass, [$resource]);
+        $decorated = $this->createStub(ResourceMetadataCollectionFactoryInterface::class);
```

Comment:
we're using `prophecy` in the rest of the code this seems inconsistent with the code base.

Reply:
PHPunit’s API is arguably better than Prophecy’s (mocks and stubs are distinguished, and they are recognized as instance of the doubled class eg.) and it does not require an additional dependency.

Do we really want to stick with Prophecy?

Reply:
agreed but yes for consistency accross the codebase